src/main/java/net/minecraftforge/registries/RegistryObject.java

```diff
@@ -474,26 +459,93 @@ public <X extends Throwable> T orElseThrow(Supplier<? extends X> exceptionSuppli
      * The returned optional will be empty if the registry does not exist or if {@link #isPresent() returns false}.
      *
      * @return an optional {@link Holder} instance pointing to this RegistryObject's name and value
+     * @deprecated Use the native methods, as this class now implements Holder.
      */
-    @NotNull
     public Optional<Holder<T>> getHolder()
     {
-        return Optional.ofNullable(this.holder);
+        return Optional.of(this);
     }
 
     @Override
     public boolean equals(Object obj)
     {
         if (this == obj) return true;
         if (obj instanceof RegistryObject) {
-            return Objects.equals(((RegistryObject<?>)obj).name, name);
+            return ((RegistryObject<?>) obj).key == this.key;
         }
         return false;
     }
 
     @Override
     public int hashCode()
     {
-        return Objects.hashCode(name);
+        return this.key.hashCode();
+    }
+
+    @Override
+    public T value()
+    {
+        return get();
+    }
+
+    @Override
+    public boolean isBound()
+    {
+        return isPresent() && this.holder.isBound();
+    }
+
+    @Override
+    public boolean is(ResourceLocation id)
+    {
+        return id.equals(this.key.location());
+    }
+
+    @Override
+    public boolean is(ResourceKey<T> key)
+    {
+        return key == this.key;
+    }
+
+    @Override
+    @SuppressWarnings({"rawtypes","unchecked"})
+    public boolean is(Predicate<ResourceKey<T>> filter)
+    {
+        return filter.test((ResourceKey) this.key);
```

Comment:
Why is this unchecked cast necessary? It looks like `this.key` is already a `ResourceKey<T>` which is what the predicate expects

Reply:
same reason as https://github.com/neoforged/NeoForge/pull/35#discussion_r1268548506